Add table relationships to use cascading deletes

johnhenley opened the pull request:
Data not deleted from tables when deleting module instance.
Add table relationships to use cascading deletes.

<!-- 
Explain the benefit of this pull request
You can erase any parts of this template not applicable to your Pull Request. 
-->

### Description of PR...

## Changes made
- Add relationships to Modules table where possible to delete cascade
- Add ModuleId to activeforums_Content table to support cascading deletes
- Add cascading deletes to various tables within the Forums module 

![image](https://user-images.githubusercontent.com/9553126/215609713-73019966-d106-4d9e-88f3-50f686619671.png)

## PR Template Checklist

- [X] Fixes Bug
- [X] Feature solution
- [ ] Other


## Please mark which issue is solved
<!-- Type numbers directly after the #, it will show the issues with that number -->

Close #97

WillStrohl commented:
Nice one!

johnhenley commented:
Since cascading delete puts the logic into the database, should the delete sprocs be revisited?

Also, any opinions on DAL2/petapoco? It looks like some objects/tables were added to use it, but never implemented?

johnhenley commented:
> > Since cascading delete puts the logic into the database, should the delete sprocs be revisited?
> 
> > 
> 
> > Also, any opinions on DAL2/petapoco? It looks like some objects/tables were added to use it, but never implemented?
> 
> 
> 
> Honestly, since DAL2 is a supported feature and would need a migration path if anything happened to it, I think we should focus on that instead of bypassing it to go directly to Petapoco (if that's what you're saying).  It would be difficult in some cases for complex queries, but there are a lot of ways to work around that.  I have many examples of this in an example code camp module I build a long time ago for a DNN Corp contest.  
> 
> 
> 
> [Example: Loading a main object in DAL2, then loading a property with related data.](https://github.com/WillStrohl/dnnextensions/blob/development/Modules/CodeCamp/Controllers/SpeakerInfoController.cs#L96-L102)  
> 
> 
> 
Actually my question should probably have been more like "this module was built before the time of DAL2, and doesn't (really) have the more modern architecture of entities (and properties with attributes) and controllers as exhibited by your code camp example and other dnn community modules. Should we be embracing that style of architecture going forward and away from a sproc for each and every database interaction?"

WillStrohl commented:
Ah... That makes more sense.  Gotcha.  

Yes, that's an excellent question.  Since you've been in there a lot more than the rest of us recently, you will likely have a better idea, but here's how I would approach that decision.  

If there is a TON of work to do to migrate all current DAL to DAL2, I think we should delay that effort for now.  That is, unless you or someone else is so passionate about it that you're willing just do it all.  Overall, this is indeed a big chunk of technical debt that we'll eventually need to get rid of.  If for no other reason, than to be able to upgrade DNN in the future.  

However, in the meantime, I think it's a realistic approach to maybe only implement DAL2 for new code that goes in.  This has some pros/cons as well though.  For example, there will be times when that isn't possible, and we'll just be adding more technical debt.

johnhenley commented:
> Ah... That makes more sense.  Gotcha.  
> 
> 
> 
> Yes, that's an excellent question.  Since you've been in there a lot more than the rest of us recently, you will likely have a better idea, but here's how I would approach that decision.  
> 
> 
> 
> If there is a TON of work to do to migrate all current DAL to DAL2, I think we should delay that effort for now.  That is, unless you or someone else is so passionate about it that you're willing just do it all.  Overall, this is indeed a big chunk of technical debt that we'll eventually need to get rid of.  If for no other reason, than to be able to upgrade DNN in the future.  
> 
> 
> 
> However, in the meantime, I think it's a realistic approach to maybe only implement DAL2 for new code that goes in.  This has some pros/cons as well though.  For example, there will be times when that isn't possible, and we'll just be adding more technical debt.  

I took a quick stab at this, separating some entities and controllers, standardizing method names (e.g. Get() rather than ForumGet(), etc.), and reorganizing namespaces a bit. It's actually much easier with refactoring tools in visual studio. 

However, there are two caveats. 

First, it will create huge commits/pull requests due to the number of files touched. Is that acceptable as part of a review process?

Second, it breaks public methods. My question would be "is anyone actually integrating with the forums module at the DLL/method level?" The only one I am aware of was engage:publish, but that is long gone. 

The forums module was originally a VB.net project which was ported in situ to c# and has become quite brittle. 

I personally think it's a worthwhile task, moving the code base forward into a better spot where we can more aggressively add features and take advantage of better practices like DAL2, web APIs, and language features like generics for instance. 

Thoughts? Should we open a discussion topic on who is integrating and if anyone, how? Should I take this as a different branch or fork and see where it goes?

WillStrohl commented:
> I took a quick stab at this, separating some entities and controllers, standardizing method names (e.g. Get() rather than ForumGet(), etc.), and reorganizing namespaces a bit. It's actually much easier with refactoring tools in visual studio.
> 
> However, there are two caveats.
> 
> First, it will create huge commits/pull requests due to the number of files touched. Is that acceptable as part of a review process?
>  

This will be inevitable once we do this kind of update.  If we can make sure the commits are separated by purpose, it will make things easier.  For example, a first pass might be white space and/or formatting differences.  Then, the next might be adding replacement methods and marking others as deprecated.  Then, the final commit with any actual changes.  Does that make sense?  If the commits are all at once, it will be  VERY time-consuming to review.  :) 

> 
> Second, it breaks public methods. My question would be "is anyone actually integrating with the forums module at the DLL/method level?" The only one I am aware of was engage:publish, but that is long gone.
> 

This is not something that is ever able to be forecasted.  Our community has long operated under the assumption that public methods are API-friendly.  So, we'll want to deprecate any (now) legacy methods/signatures for removal at least 1-2 major or minor versions out.  We don't have an official policy on this module, specifically, but it would make sense to try and mimic the core of DNN.  

>  
> The forums module was originally a VB.net project which was ported in situ to c# and has become quite brittle.
> 
> I personally think it's a worthwhile task, moving the code base forward into a better spot where we can more aggressively add features and take advantage of better practices like DAL2, web APIs, and language features like generics for instance.
> 
> Thoughts? Should we open a discussion topic on who is integrating and if anyone, how? Should I take this as a different branch or fork and see where it goes?

I think this is a great idea!  Maybe it might spark some interest and excitement about that and other issues we need to get help with.  :)

johnhenley commented:
What if we did something like this, where we facade and inherit everything from the "new" code?
![image](https://user-images.githubusercontent.com/9553126/220764456-728ebf19-9d98-4ce8-a61b-9b3ecbf954c1.png)

WillStrohl commented:
Honestly, I love that idea... Especially since it appears to solve both issues: (1) we can safely deprecate methods and notify developers; and (2) it helps us move to a better architecture.

johnhenley commented:
> Honestly, I love that idea... Especially since it appears to solve both issues: (1) we can safely deprecate methods and notify developers; and (2) it helps us move to a better architecture.

@WillStrohl I created a small PR #251 for the "Likes" and "LikesController" classes to see if this is workable very small units of work. Let me know what you think....

johnhenley commented:
When installing this update to an instance where there was a forums module that was deleted and left (orphaned) data behind, the SQL to add the constraints fails. There are two or three options I think:
- delete any orphaned data during the install of the upgrade 
- create a fake module record to satisfy integrity 
- provide instructions in the release notes to run some queries to determine if you have orphaned data 
Neither probably ideal but needs to be addressed since this will undoubtedly be a problem.

johnhenley commented:
> When installing this update to an instance where there was a forums module that was deleted and left (orphaned) data behind, the SQL to add the constraints fails. There are two or three options I think:
> 
> * delete any orphaned data during the install of the upgrade
> * create a fake module record to satisfy integrity
> * provide instructions in the release notes to run some queries to determine if you have orphaned data
>   Neither probably ideal but needs to be addressed since this will undoubtedly be a problem.

1. Updated the upgrade SQL to add a fake module record with IsDeleted = 1 to satisfy the foreign key constraints that reference the Modules table
2. Added some logic to remove orphaned topic tracking and content attachment records. I found that in my testing against my own data that sometimes topics were deleted [probably manually] but some referencing rows were not properly deleted.
3. Added some logic during the upgrade process to compare the attachment folder(s) against the attachment tables in the database and move any orphaned files (that don't have a corresponding row in the attachments table) to an "orphaned" subfolder.

WillStrohl commented on components/Topics/TopicsController.cs:
```diff
@@ -534,10 +534,66 @@ public override IList<SearchDocument> GetModifiedSearchDocuments(ModuleInfo modu
+                    }
+
+                    break;
+                case "08.00.00":
```
Hey @johnhenley ...  I don't think this will have any adverse impact, but it is referencing 8.0 code.  Do you know if there is anything underlying that we should be worried about before I process this?

johnhenley replied:
The 8.0 case needs to be removed because it's probably not going to compile. Let me know if you want me to do a pull request.

WillStrohl replied:
Oh, then maybe the commit would be best?  I was about to try myself.

johnhenley replied:
You should be able to commit now as I resolved the conflicts. Not sure what happened but looks ok now

WillStrohl reviewed (approved):
Looks good!
